pkg/kinds/general.go

```diff
@@ -152,17 +152,211 @@ func (m *GrafanaResourceMetadata) GetOriginInfo() *ResourceOriginInfo {
 	if !ok {
 		return nil
 	}
-	info := &ResourceOriginInfo{
-		Name: v,
-		Path: m.Annotations[annoKeyOriginPath],
-		Key:  m.Annotations[annoKeyOriginKey],
+	return &ResourceOriginInfo{
+		Name:      v,
+		Path:      m.GetOriginPath(),
+		Key:       m.GetOriginKey(),
+		Timestamp: m.GetOriginTimestamp(),
 	}
-	v, ok = m.Annotations[annoKeyOriginTimestamp]
-	if ok {
-		t, err := time.Parse(time.RFC3339, v)
-		if err != nil {
-			info.Timestamp = &t
+}
+
+func (m *GrafanaResourceMetadata) GetOriginName() string {
+	return m.Annotations[annoKeyOriginName]
+}
+
+func (m *GrafanaResourceMetadata) GetOriginPath() string {
+	return m.Annotations[annoKeyOriginPath]
+}
+
+func (m *GrafanaResourceMetadata) GetOriginKey() string {
+	return m.Annotations[annoKeyOriginKey]
+}
+
+func (m *GrafanaResourceMetadata) GetOriginTimestamp() *time.Time {
+	v, ok := m.Annotations[annoKeyOriginTimestamp]
+	if !ok {
+		return nil
+	}
+	t, err := time.Parse(time.RFC3339, v)
+	if err != nil {
+		return nil
+	}
+	return &t
+}
+
+// Accessor functions for k8s objects
+type GrafanaResourceMetaAccessor interface {
+	GetUpdatedTimestamp() *time.Time
+	SetUpdatedTimestamp(v *time.Time)
+	GetCreatedBy() string
+	SetCreatedBy(user string)
+	GetUpdatedBy() string
+	SetUpdatedBy(user string)
+	GetFolder() string
+	SetFolder(uid string)
+	GetSlug() string
+	SetSlug(v string)
+	GetOriginInfo() *ResourceOriginInfo
+	SetOriginInfo(info *ResourceOriginInfo)
+	GetOriginName() string
+	GetOriginPath() string
+	GetOriginKey() string
+	GetOriginTimestamp() *time.Time
+}
+
+var _ GrafanaResourceMetaAccessor = (*grafanaResourceMetaAccessor)(nil)
+var _ GrafanaResourceMetaAccessor = (*GrafanaResourceMetadata)(nil)
+
+type grafanaResourceMetaAccessor struct {
+	obj v1.Object
+}
+
+func MetaAccessor(obj v1.Object) GrafanaResourceMetaAccessor {
+	return &grafanaResourceMetaAccessor{obj}
+}
+
+func (m *grafanaResourceMetaAccessor) set(key string, val string) {
+	anno := m.obj.GetAnnotations()
+	if val == "" {
+		if anno != nil {
+			delete(anno, key)
 		}
+	} else {
+		if anno == nil {
+			anno = make(map[string]string)
+		}
+		anno[key] = val
+	}
+	m.obj.SetAnnotations(anno)
```

Comment:
^^^ this seems to be required for unstructured to work 🤷🏻, for other implementations, it may not be required